static/js/reactions.js

```diff
@@ -97,16 +86,52 @@ exports.toggle_emoji_reaction = function (message_id, emoji_name) {
         return;
     }
 
-    var has_reacted = exports.current_user_has_reacted_to_emoji(message, emoji_name);
+    var reaction_info = {
+        emoji_name: emoji_name,
+    };
+
+    if (emoji.all_realm_emojis.hasOwnProperty(emoji_name)) {
+        if (emoji_name === 'zulip') {
+            reaction_info.reaction_type = 'zulip_extra_emoji';
+        } else {
+            reaction_info.reaction_type = 'realm_emoji';
+        }
+        reaction_info.emoji_code = emoji_name;
+    } else if (emoji_codes.name_to_codepoint.hasOwnProperty(emoji_name)) {
+        reaction_info.reaction_type = 'unicode_emoji';
+        reaction_info.emoji_code = emoji_codes.name_to_codepoint[emoji_name];
+    } else {
+        blueslip.warn('Bad emoji name: ' + emoji_name);
+        return;
+    }
+
+    var has_reacted = exports.current_user_has_reacted_to_emoji(
+        message,
+        reaction_info.emoji_name,
+        reaction_info.reaction_type
+    );
     var operation = has_reacted ? 'remove' : 'add';
 
-    send_reaction_ajax(message_id, emoji_name, operation);
+    send_reaction_ajax(message_id, reaction_info, operation);
 
     // The next line isn't always necessary, but it is harmless/quick
     // when no popovers are there.
     emoji_picker.hide_emoji_popover();
 };
 
+exports.process_reaction_click = function (message_id, local_id) {
+    var reaction_info = exports.get_reaction_info(local_id);
+    var message = get_message(message_id);
+    var has_reacted = exports.current_user_has_reacted_to_emoji(
+        message,
+        reaction_info.emoji_name,
+        reaction_info.reaction_type
+    );
+    var operation = has_reacted ? 'remove' : 'add';
+
+    send_reaction_ajax(message_id, reaction_info, operation);
+};
```

Comment:
It feels like this basically duplicates the buttom third of the previous function; is that avoidable?